File: Problem2.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No
class Solution {
    Map<Character,Integer> map;
    public boolean isAlienSorted(String[] words, String order) {
        this.map = new HashMap<>();
        for(int i = 0; i < order.length(); i++){
            char ch = order.charAt(i);
            map.put(ch,i);
        }
        for(int i = 0; i < words.length - 1; i++){
            String word1 = words[i];
            String word2 = words[i+1];
            if(notOrder(word1,word2))
                return false;
        }
        return true;
    }
    public boolean notOrder(String word1, String word2){
        int i = 0;
       
        for(i = 0; i < word1.length() && i < word2.length(); i++){
            char c1 = word1.charAt(i);
            char c2 = word2.charAt(i);
            if(c1 != c2){
                if(map.get(c1) > map.get(c2))
                    return true;
                    break;
            }
        }
        if(i!=word1.length() && i == word2.length())
            return true;
        return false;
    }
}

File: problem2.cpp
/*
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
create hashmap for the order. then compare consecutively if smaller continue onto next word
else if larger return false
if consecutively true then return true;
*/

#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
    vector<int> hash_map{};
    bool foo(string& s1,string& s2){
        int l1 = s1.size();
        int l2 = s2.size();
        for(int i{};i<l1 && i<l2;++i){
            //cout<<hash_map.at(s1.at(i)-97)<<" "<<hash_map.at(s2.at(i)-97)<<endl;
            int v1 = hash_map.at(s1.at(i)-97);
            int v2 = hash_map.at(s2.at(i)-97);
            if(v1 == v2){
                continue;
            }
            else if(v1<v2){
                return true;
            }
            else{
                return false;
            } 
        }
        if(l1>l2){
            return false;
        }
        return true;
    }
    void display(vector<int>& v1){
        for(int& v:v1){
            cout<<v<<" ";
        }
        cout<<endl;
    }
public:
    bool isAlienSorted(vector<string>& words, string order) {
        int word_length = words.size();
        if (word_length < 2) return true;
        int sz = order.length();
        hash_map.resize(sz,0);
        for(int i{};i<sz;++i){
            hash_map.at(order.at(i)-97) = i;
        }
        //display(hash_map);
        for(int i{};i<word_length-1;++i){
            if(foo(words.at(i),words.at(i+1)) == false){
                return false;
            }
        }
        return true;
    }
};